examples/hubert/dataset/dataset.py

```diff
@@ -0,0 +1,214 @@
+            )
+        self.data_source = data_source
+        self.max_token_count = max_token_count
+        self.buckets = self._get_buckets(self.data_source, num_buckets)
+
+    def _get_buckets(
+        self,
+        data_source: Dataset,
+        num_buckets: int
+    ) -> List[int]:
+        buckets = {}
+        len_list = data_source.len_list
+        min_len, max_len = min(len_list), max(len_list)
+
+        boundries = [min_len - 1]
+        interval = (max_len - min_len) // num_buckets
+        for i in range(1, num_buckets):
+            boundries.append(min_len + i * interval)
+        boundries.append(max_len + 1)
+        bucket_ids = torch.bucketize(torch.Tensor(len_list), torch.Tensor(boundries))
+        for i, length in enumerate(len_list):
+            bucket_id = bucket_ids[i]
+            if bucket_id in buckets:
+                buckets[bucket_id].append(i)
+            else:
+                buckets[bucket_id] = [i]
+        for k in buckets:
+            random.shuffle(buckets[k])
+        return buckets
+
+    def __iter__(self) -> Iterator[List[int]]:
+        iter_list = []
+        total_len = 0
+        batch = []
+        len_list = self.data_source.len_list
+        if self.max_token_count:
+            for k in self.buckets.keys():
+                for index in self.buckets[k]:
+                    if total_len + len_list[index] > self.max_token_count:
+                        iter_list.append(batch)
+                        batch = [index]
+                        total_len = len_list[index]
+                    else:
+                        batch.append(index)
+                        total_len += len_list[index]
+        else:
+            for index in self.buckets[k]:
+                if len(batch) == self.batch_size:
+                    iter_list.append(batch)
+                    batch = [index]
+                    total_len = len_list[index]
+                else:
+                    batch.append(index)
+                    total_len += len_list[index]
+
+        for batch in iter_list:
+            yield batch
+
+    def __len__(self):
+        return len(self.data_source)
+
+
+class HuBERTDataSet(Dataset):
+    """Create a Dataset for HuBERT model training and fine-tuning.
+
+    Args:
+        tsv_dir (str or Path): The root directory of the ``.tsv`` file list.
+        phase_type (str): The type of the dataset. Options: [``train``, ``valid``, ``test``].
+    """
+    def __init__(
+        self,
+        exp_dir: Union[str, Path],
+        dataset: str,
+        phase_type: str,
+        min_sample: int = 32000,
+        max_sample: int = 250000,
+    ) -> None:
+        self.exp_dir = Path(exp_dir)
+        tsv_dir = self.exp_dir / "tsv"
+        label_dir = self.exp_dir / "label"
+        f_list, ind_list, len_list = self._get_lists(
+            tsv_dir,
+            dataset,
+            phase_type,
+            min_sample,
+            max_sample
+        )
```

Comment:
I am wondering about the size of these returned lists? Just want to give a heads up that multiprocessing with python list would cause N times memory usage. N represents the number of workers.

Using NumPy array or Tensor would help sharing data across processes without copy.

Reply:
Good point. The size is actually quite large here. I will refactor it by using Tensor.